src/webcore/promise_future.rs

```diff
@@ -83,6 +84,17 @@ impl PromiseFuture< (), Never > {
         B: Future< Item = (), Error = Never > + 'static {
         spawn_local( future );
     }
+
+    /// This is used to print an error to the console.
+    ///
+    /// See the documentation for [`spawn_local`](#method.spawn_local) for more details.
+    #[inline]
+    pub fn print_error< A: JsSerialize >( value: A ) -> Never {
```

Comment:
I feel like the naming of this method is slightly misleading, as it doesn't just print the error - it also panics. Perhaps simply calling it `panic` would be better?

Reply:
I think the `Never` return type is good enough to indicate that it panics/never returns.

Calling it `panic` is wrong because it doesn't just panic (and in fact panicking is not its primary purpose).

Reply:
But you don't see the `Never` type when reading code, you see `.map_err(PromiseFuture::print_error)` and it's not obvious that it's going to panic. And the `Never` type (or `!`) for that matter *in theory* can also mean that it goes into an infinite loop.

The user should be always fully aware that their code can panic, especially because currently panics can cause issues like memory leaks and `RefCell`s getting stuck in a borrowed state.

Reply:
> But you don't see the Never type when reading code, you see .map_err(PromiseFuture::print_error) and it's not obvious that it's going to panic.

Indeed, though that's just how Futures work: they require `Never` for the error type, so if you're using `map_err` it's pretty safe to assume it panics. So anybody with familiarity with Futures 0.2 should understand what's going on.

> The user should be always fully aware that their code can panic

A good principle, yes. Though in both the standard library *and* stdweb that isn't enforced *by the name*, instead it's usually done by adding in a `# Panics` section to the docs, which I will add.

In any case, `panic` alone is a really bad name. `error_panic` isn't good either. `print_error_panic` is okay, but quite long. Maybe it's acceptable.

Reply:
> especially because currently panics can cause issues like memory leaks and RefCells getting stuck in a borrowed state.

That doesn't sound right. Right now in wasm panic = abort, so that won't ever happen.

And even with normal panics I can't think of any way to abuse panics like that: when a panic happens, it drops everything in scope.

Do you have any examples (even theoretical/hypothetical ones)?

Reply:
> In any case, panic alone is a really bad name. error_panic isn't good either. print_error_panic is okay, but quite long. Maybe it's acceptable.

Or maybe `print_and_panic`?

> That doesn't sound right. Right now in wasm panic = abort, so that won't ever happen. And even with normal panics I can't think of any way to abuse panics like that: when a panic happens, it drops everything in scope.

A panic will, not drop anything and generate a `RuntimeException` on the JS side. On other targets this is not a problem since an abort will kill the problem, but here there is nothing preventing the JS code from calling into the Rust code again.

Reply:
E.g. a program like this:

```
#[macro_use]
extern crate stdweb;

struct Dummy;
impl Drop for Dummy {
    fn drop( &mut self ) {
        console!( log, "Drop called" );
    }
}

fn main() {
    let dummy = Dummy;
    panic!("");
}
```

On emscripten targets the object will be properly dropped, but on `wasm32-unknown-unknown` it will not.

Reply:
Hmmm, you're right that JS can call back into wasm again, I hadn't thought of that.